dhall/src/Dhall/Repl.hs

```diff
@@ -260,7 +264,108 @@ hashBinding tokens = do
   liftIO . Text.putStrLn $
     hashExpressionToCode _standardVersion normalizedExpression
 
+saveFilePrefix :: FilePath
+saveFilePrefix = ".dhall-repl"
+
+-- | Find the index for the current _active_ dhall save file
+currentSaveFileIndex :: MonadIO m => m (Maybe Int)
+currentSaveFileIndex = do
+  files <- liftIO $ getDirectoryContents "."
+
+  let parseIndex file
+        | saveFilePrefix `isPrefixOf` file
+        , '-':index <- drop (length saveFilePrefix) file
+        , [(x, "")] <- reads index -- safe version of read
+        = Just x
+
+        | otherwise
+        = Nothing
+
+  pure $ case mapMaybe parseIndex files of
+    [] -> Nothing
+    xs -> Just $ maximum xs
+
+-- | Find the name for the current _active_ dhall save file
+currentSaveFile :: MonadIO m => m (Maybe FilePath)
+currentSaveFile =
+  (fmap . fmap) (\i -> saveFilePrefix <> "-" <> show i) currentSaveFileIndex
+
+-- | Find the name for the next dhall save file
+nextSaveFile :: MonadIO m => m FilePath
+nextSaveFile = do
+  mIndex <- currentSaveFileIndex
+
+  let nextIndex = maybe 0 succ mIndex
+
+  pure $ saveFilePrefix <> "-" <> show nextIndex
+
+loadBinding :: ( MonadIO m, MonadState Env m ) => [String] -> m ()
+loadBinding [] = do
+  mFile <- currentSaveFile
+
+  case mFile of
+    Just file -> loadBinding [file]
+    Nothing   ->
+      fail $ ":load couldn't find any `" <> saveFilePrefix <> "-*` files"
+
+loadBinding [file] = do
+  loadedExpression <- parseAndLoad =<< liftIO (readFile file)
+
+  -- Typecheck and normalize to support modified save files
+  _ <- typeCheck loadedExpression
+  normalizedExpression <- normalize loadedExpression
+
+  case normalizedExpression of
+    Expr.RecordLit fields -> do
+
+      bindings <- forM (Map.toMap fields) $ \expr -> do
+        type' <- typeCheck expr
+        pure $ Binding { bindingType = type', bindingExpr = expr }
+
+      modify
+        ( \e ->
+            e
+              { envBindings =
+                  Data.Map.foldrWithKey Dhall.Context.insert (envBindings e) bindings
+              }
+        )
+
+      liftIO . putStrLn $ "\nLoaded `" <> file <> "`\n"
+
+    _  -> fail ":load expects a file that contains a record"
+
+loadBinding _ = fail ":load should be of the form `:load` or `:load file`"
+
 saveBinding :: ( MonadIO m, MonadState Env m ) => [String] -> m ()
+-- Save all the bindings into a big record
+saveBinding [] = do
+  file <- nextSaveFile
+
+  saveBinding [file]
+
+-- Save all the bindings into a big record into `file`
```

Comment:
We probably should use something other than a record to store the bindings because we need to preserve the original order.  The context permits duplicate entries and if their order is switched then you change which one shadows which

Reply:
I saw that the context keeps the shadowed bindings, but I'm not too sure why.

Maybe if we weren't normalizing the bindings before adding them to the context then we would need to have access to the shadowed bindings. But I believe that when we normalise while adding a new binding, it gets rid of any dependencies on previous bindings, or is that not the case?

Or maybe we need access to the shadowed bindings for something else?

The current solutions I have is to only serialise the only accessible version of a binding accessible to the user when they type `:save`.

Reply:
@basile-henry: It's still possible to refer to shadowed bindings using the `@n` suffix, like this:

```
⊢ :let x = 1

x : Natural

⊢ :let x = True

x : Bool

⊢ x@1

1
```

Reply:
Ah, I didn't know about this feature.

I liked the idea of using dhall to serialise the REPL state. What would you recommend as a format?

I could use `[ { name : Text, expr : Text } ]` but then the save file can't really be used for anything else.